Yulia/app data update

vveiln opened the pull request:
This PR adds the following changes to the spec:
- a new proposed `applicationData` structure with four reserved fields: resourcePayload, discoveryPayload, externalPayload, and logicVKOuter hash.
- replaces the previous `logicVKOuter` field in `logicVerifierInputs` with a generic `verifyingKey` and moves `logicVKOuter` to `appData`, as described above
- removes the `isConsumed` flag from `logicVerifierInputs` as in can be inferred from the compliance instance structure

⚠️ The scope of this PR is the three changes described above. All other requested changes are out of the scope ⚠️

All relevant changes are in this file: docs/arch/system/state/resource_machine/data_structures/action/index.juvix.md (everything else is some random stuff git made me do)

jonaprieto committed:
chore: update version to 1.0.0 and remove deprecated documentation

- Updated project version to 1.0.0 in pyproject.toml and VERSION file.
- Removed obsolete Juvix documentation files related to various node engines and subsystems to streamline the codebase.
- Adjusted mkdocs.yml to reflect the changes in documentation structure.

jonaprieto committed:
chore(docs): update architecture and tag details

Simplified the architecture section by removing the Node architecture details to reference it as a work-in-progress found in release v0.2.0. Updated tags in multiple documentation files, aligning them with the current scope of content and removing outdated references to the Node architecture. These changes aim to streamline the documentation by focusing on existing, fully developed components while acknowledging ongoing development work separately.

jonaprieto committed:
docs(changelog): update with node architecture removal note

Added a note about the removal of node architecture in the latest release. The node architecture details are available in the v0.2.0 documentation, with plans to reintegrate it in the upcoming v2.0.0 release.

vveiln committed:
draft appdata proposal

vveiln committed:
Add more application-data-related descriptions

vveiln committed:
Update verifying keys stuff; remove isConsumed flag

vveiln committed:
Minor fix

XuyangSong reviewed (approved):
lgtm
Left a comment regarding the commitment to `logicRef` in `logicVKOuter`.

agureev reviewed (commented):
Minor thought in addition: what if the type of appdata will instead be:

`Map(payload => List(Bitsring, DeletionCriterion))`

where `payload` can be any of some encoding of strings `resourcePayload`, `discoveryPayload`, `externalCallPayload`, `logicVKOuter`, `appPayload`

That way "empty field" comment would make sense to me as that would mean that the key would just have an empty list attached.

We can then just decide which payload goes in which order for the circuit to process.

vveiln commented on docs/arch/system/state/resource_machine/data_structures/action/index.juvix.md:
```diff
@@ -16,11 +16,30 @@ An action is a composite structure of type `Action` that contains the following
+|`resourcePayload`|Contains the resource-object-related data. For example, in the shielded case, it contains encrypted resource object. If no resource payload is expected to be verified, the field is left empty.|
+|`discoveryPayload`|Contains the discovery-related data, for example, FMD ciphertext. If no discovery payload is expected to be verified, the field is left empty.|
+|`externalCallPayload`|Contains the data associated with external calls. If no external calls are made or no external calldata has to be verified, the field is left empty.|
+|`logicVKOuter`|In the DP (data privacy only) case, contains `logicRef` associated with the corresponding resource. In the FP (data and function privacy) case, contains a hiding and binding commitment to `logicRef`. It can be said that in the DP case outer hash is an identity function and in the FP case it is instantiated by a hiding and binding commitment scheme.|
```
I'll remove it

vveiln commented:
> That way application circuits to not have to be explicitly aware of any additional arguments, e.g. the trivial logic previously did not need anything in appdata to generate proofs, now it has all these empty hashes fed to it inside the circuit.

The verifier would know that an empty hash is an empty parameter and just wouldn't pass it to the logic, the logic wouldn't have to see these inputs.

I agree with the encoding point though, it also would offload the work from the circuit, which is always good.

I'll change the format to the proposed one.

agureev commented:
@vveiln another idea I just had is maybe just make it a 5-tuple of lists? So the type would be `{List(BlobWIthDelCriterion), List(BlobWIthDelCriterion), List(BlobWIthDelCriterion), List(BlobWIthDelCriterion), List(BlobWIthDelCriterion)}`

Where each position is beforehand tied to one of the payloads?

That way we don't have to worry about how to encode names of the fields and beforehand decide on the order all of this goes in to the instance

vveiln committed:
Remove logicVKOuter from application data; change the type of application data

vveiln commented:
I removed `logicVKOuter` from `applicationData` and changed `applicationData` type to:
```
(
  [(BitString, DC), ..] #resourcePayload
  [(BitString, DC), ..] #discoveryPayload
  [(BitString, DC), ..] #externalPayload
  [(BitString, DC), ..] #applicationPayload
)
```

jonaprieto reviewed (changes requested):
Why are you deleting all those files? See the files changed

vveiln commented:
These changes were generated by git automatically. As I understand, this is the diff between main and v1.0.0 – all the files related to engines and such removed, but I might be wrong.

XuyangSong commented:
> I removed `logicVKOuter` from `applicationData` and changed `applicationData` type to:
> 
> ```
> (
>   [(BitString, DC), ..] #resourcePayload
>   [(BitString, DC), ..] #discoveryPayload
>   [(BitString, DC), ..] #externalPayload
>   [(BitString, DC), ..] #applicationPayload
> )
> ```

Are we using a fixed array here? The struct would look like this: 
```rust
struct AppData ( [Vec<(Vec<u8>, DC)>; 4] )
```

If only four entries are involved, do you think we can add the payload names to the struct? It seems better than relying on fixed positions. such as
```rust
struct AppData {
    resource_payload: Vec<(Vec<u8>, DC)>,
    discovery_payload: Vec<(Vec<u8>, DC)>,
    external_payload: Vec<(Vec<u8>, DC)>,
    application_payload: Vec<(Vec<u8>, DC)>,
}
```


If we plan to extend payloads to five or six in the future, the fixed array may not be suitable, and we can use a variable array instead, like:
```rust
struct AppData (Vec<Vec<(Vec<u8>, DC)>>)
```
Then developers need to remember the positions.

@vveiln @agureev 
Let me know your preference.

agureev commented:
> If only four entries are involved, do you think we can add the payload names to the struct?

I think this can be seen as an implementation detail. The struct you have can have the same semantics as the 4-tuple. Either way we have to order them in some way in order to know in which order to place them into the instance.

In other words, I think we can make the appdata the struct you mentioned without needing to change the current spec. So I am ambivalent as to whether we change it to an ADT or just keep it like a tuple.

But I actually would also like for the implementation to be a stuct on the Rust/Solidity side.

vveiln committed:
Add action tree size and minimal description; add applicationData processing instructions

vveiln committed:
Merge branch 'v1.0.0' into yulia/app-data-update

agureev commented on docs/arch/system/state/resource_machine/data_structures/action/index.juvix.md:
```diff
@@ -18,6 +18,7 @@ An action is a composite structure of type `Action` that contains the following
 |-|-|-|
 |`logicVerifierInputs`|`Map Tag LogicVerifierInputs`|For each resource tag, contains the associated logic proof and everything required to verify it. The structure of `LogicVerifierInputs` is further described below.|
 |`complianceUnits`|`List ComplianceUnit`|The set of transaction's [[Compliance unit | compliance units]]|
+|`actionTreeSize`||Determines the depth of the Merkle tree used to store input resources. If not specified, the depth is set to the smallest necessary to fit all the input resources.|
```
Should this be part of the action? Why do we need to let the user provide it?

Wouldn't it be the RM that decides beforehand whether it will use the smallest action tree needed or a fixed-size one based on the proving system it uses?

vveiln replied:
Because we want the verifier to know the depth of the tree to build in case the depth is not set dynamically. It is added for compatibility with proving systems that do not support variable tree size like zkvms do.

agureev replied:
I have to think about that.

But also: what is the type of the input and what do we mean by "empty"?

Maybe we can decide on a type of integers and if the value is -1 we construct the least size one?

Or maybe we can just make it natural numbers? i.e. not include the option for it to be "empty"?

That way it's the higher-level interface which will compute that number if needed.

agureev replied:
I do think I now understand the usefulness of this for RMs that have proving systems that can only have fixed-size argments and hence support only fixed-size merkle proofs.

This way it's the application developer who decides how many resources can exist with the resource in the same action instead of us. E.g. currently we support only 16 resources per action, it is a cutoff based on the RM design as that's the tree which we construct

But now you submit the tree depth information yourself and hence determine the proof length that you expect

However, that does mean only apps which expect the same depth can be part of the same action

So every application can have different verifying keys based on which tree depth they expect

I dunno how much we would want that...

But for the veritable tree depth supporting proving systems, how does it help? I am unsure still

agureev commented on docs/arch/system/state/resource_machine/data_structures/action/resource_logic_proof.juvix.md:
```diff
@@ -14,25 +14,31 @@ module arch.system.state.resource_machine.data_structures.action.resource_logic_
+
+The original order of the elements must be preserved at each step.
 
 #### Witness
```
Outside of this PR but wanted to mention:

At some point we should probably change this page to reflect the part that we don't control the witness of the resource logics.

Moreover, parts 3 and 4 seem to imply that we put all the resources created and consumed into the witness info. However as we have discussed this actually rarely happens and we don't even have apps that have such witness structure.

vveiln replied:
> At some point we should probably change this page to reflect the part that we don't control the witness of the resource logics.

I'm not sure what you mean by that. We generally can't control anything that is provided by the user, but we specify what we expect to receive

> Moreover, parts 3 and 4 seem to imply that we put all the resources created and consumed into the witness info. However as we have discussed this actually rarely happens and we don't even have apps that have such witness structure.

I'll add a note that passing all resource objects as witnesses is not necessary, but we can't say if it happens rarely in principle

agureev reviewed (commented):
Sorry for vetoing the upgrade again, but I am unsure why we are putting the depth in the Action datatype. Could you elaborate on that?

Also, do we need to change anything for the Compliance Unit? It would be really nice if the compliance unit could accept variable-size commitment-accumulator.
